### PR TITLE
Add auto-parallelism to Jobs (0.8 branch)

### PR DESCRIPTION
**Pull request by StephanEwen**

### CODE REVIEW

**Comment by mxm** [comment]
Looks good. That way, we can help users by preventing them from running programs with the default degree of parallelism (=1) if more task slots are available.

**Comment by rmetzger** [comment]
Cool.
Lets merge this also to master and document it there.

**Comment by tillrohrmann** [comment]
The user has to enable the auto parallelism explicitly, right?

**Comment by mxm** [comment]
Right now, the user has to set the parallelism to `ExecutionConfig.PARALLELISM_AUTO_MAX`. Why not use all available task slots by default? I understand, that we shouldn't simply grab all resources but the auto parallelism will only grab the resources which were already granted to Flink.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobManager.java [hunk 1]
```diff
@@ -374,6 +375,8 @@ public JobSubmissionResult submitJob(JobGraph job) throws IOException {
 				LOG.debug(String.format("Running master initialization of job %s (%s)", job.getJobID(), job.getName()));
 			}
 
+			final int numSlots = scheduler.getTotalNumberOfSlots();
```
Comment (mxm) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/JobManager.java]: Shouldn't this be set to `getNumberOfAvailableSlots()` for the PARALLELISM_AUTO_MAX case?

**Comment by fhueske** [comment]
Using max parallelism basically prohibits to run more than one program at a time. I don't think that would be a good default mode.

**Comment by StephanEwen** [comment]
I agree with Fabian that it is not a good default behavior to grab everything that is possible.
It should be an explicit request by the user. For YARN single job sessions, we can make this the default, otherwise it is not very friendly.

`getNumberOfAvailableSlots()` changes very fast during multi user operation. Dusing single user operation between jobs (where I see the auto parallelism useful), it is the same as `getTotalNumberOfSlots()`.

**Comment by uce** [comment]
+1

I think setup via `ExecutionConfig` is the way to go.

I agree with @rmetzger that we should merge it to master as well. The important thing is to document it though. :-)

**Comment by tillrohrmann** [comment]
But currently the system does not support multi-user/multi-job scenarios so
well either. If I'm not mistaken, then the scheduler schedules the tasks
eagerly which means that two jobs could take required slots away from each
other. As a consequence, both will fail if not properly configured.

**Comment by rmetzger** [comment]
Are there any plans to merge this to master as well?
I need this feature to implement a testcase.

**Comment by rmetzger** [comment]
Ping ...

**Comment by mxm** [comment]
@rmetzger I don't see a reason why this should not go to master as well. After all, it's optional and quite useful if you want to run a job on the full cluster with as many available slots as possible.

**Comment by StephanEwen** [comment]
Manually merged into `release-0.8` in a6f9f9939ca03026baeefb3bd0876b90068b7682

**Comment by rmetzger** [comment]
Thank you for merging it!

**Comment by fpompermaier** [comment]
Hi to all,
I was reading this interesting thread..is there any change that multi-user/multi-job scenarios will come into play sooner or later? Or do you just rely on YARN for it?

**Comment by rmetzger** [comment]
Hey,
Flink already supports running multiple jobs in parallel.
If you have 50 slots available, you can run two jobs requiring 25 slots.
The webfrontend is not really able to properly report the status of concurrent jobs, but thats only a visualization issue.

**Comment by fpompermaier** [comment]
That's true but what if there's not enough resources? Is there any policy to retry the job submission automatically or give priority to waiting/queued ones?

**Comment by tillrohrmann** [comment]
At the moment, this is not supported yet. The easiest way to execute
multiple jobs concurrently is to start each job in a separate Flink cluster
running on YARN.

**Comment by fpompermaier** [comment]
I know that in stratosphere there was an effort to write a job scheduler, do you think that such a thing could be valuable for the future or are you going to rely only on hadoop-ecosytem stuff (like Oozie or Falcon upon YARN)?

**Comment by fhueske** [comment]
I think it would be definitely good to have something like a job submission
queue, that accepts jobs and executes them as soon as enough as enough
resource become available.
That should not be too hard to do.
Also simple dependencies could be checked like "execute job Y only if job X
successfully completed".

However, I am not aware of any effort in that direction.

2015-03-06 11:26 GMT+01:00 Flavio Pompermaier <notifications@github.com>:

**Comment by fpompermaier** [comment]
That would be awesome :)
I think you could talk with Markus about the Dopa scheduler..propably it's a closed project but it could be a source of inputs to create a ticket for contributors who wants to implement that!

**Pull request closed without merging**